File: NextGreaterII.cpp
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
//Next greater element is found using stack

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n= nums.size();
        vector<int> res(n,-1);
        stack<int> st;
        
        for(int i=0;i<2*n;i++){
            while(!st.empty() && nums[i%n]>nums[st.top()]){
                res[st.top()] = nums[i%n];
                st.pop();
            }
            if(i<n)
                st.push(i);
            
        }
        return res;
    }
};

File: nextGreaterII.py
'''
We can make use of stack here
Time Complexity --> O(n) for time
Space Complexity --> O(n) for stack
'''
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        res = [-1 for x in range(len(nums))]
        stk = []
        n = len(nums)
        for i in range(2*len(nums)):
            i = i%n
            if len(stk)==0:
                stk.append(i)
            else:
                while len(stk)!=0 and nums[i]>nums[stk[-1]]:
                    res[stk[-1]] = nums[i]
                    stk.pop(-1)
                stk.append(i)
        return res



